staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	// different vendors. It is provided by Kubernetes and matches the
+	// `driverName` from the ResourceSlice which provides the device.
+	//
+	// The CEL expression is applied to *all* available devices from any driver.
+	// The expression has to check for existence of an attribute when it is not
+	// certain that it is provided because runtime errors are not automatically
+	// treated as "don't select device". Instead, device selection fails completely
+	// and reports the error.
+	//
+	// Some more examples:
+	//
+	//    "memory" in device.capacities["dra.example.com"] && # Is the capacity available?
+	//       device.capacities["dra.example.com"].memory.isGreaterThan(quantity("1Gi")) # >= 1Gi
+	//
+	//    device.attributes["dra.example.com"].driverVersion.isGreaterThan(semver("1.0.0")) # >= v1.0.0, runtime error if not available
+	//
+	//    device.driverName == "dra.example.com" # any device from that driver
+	//
+	// +required
+	Expression string `json:"expression" protobuf:"bytes,1,name=expression"`
+}
+
+// Besides the request name slice, constraint must have exactly one field set.
+type Constraint struct {
+	// The constraint applies to devices in these requests. A single entry is okay
+	// and used when that request is for multiple devices.
+	//
+	// If empty, the constrain applies to all devices in the claim.
+	//
+	// +optional
+	// +listType=atomic
+	RequestNames []string `json:"requestNames,omitempty" protobuf:"bytes,1,opt,name=requestNames"`
+
+	// The devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "dra.example.com/numa" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
+	//
+	// +required
+	MatchAttribute *string `json:"matchAttribute,omitempty" protobuf:"bytes,2,opt,name=matchAttribute"`
+
+	// TODO (?)
+	// MatchQuantity *string
+
+	// Future extension, not part of the current design:
+	// A CEL expression which compares different devices and returns
+	// true if they match.
+	//
+	// Because it would be part of a one-of, old schedulers will not
+	// accidentally ignore this additional, for them unknown match
+	// criteria.
+	//
+	// matcher string
+}
+
+// ClaimConfiguration is used for configuration parameters in ResourcClaimSpec.
+type ClaimConfiguration struct {
+	// The configuration applies to devices in these requests.
+	//
+	// If empty, the configuration applies to all devices in the claim.
+	//
+	// +optional
+	// +listType=atomic
+	RequestNames []string `json:"requestNames,omitempty" protobuf:"bytes,1,opt,name=requestNames"`
+
+	Configuration `json:",inline" protobuf:"bytes,2,name=configuration"`
+}
+
+// Configuration must have exactly one field set. It gets embedded
+// inline in some other structs which have other fields, so field names must
+// not conflict with those.
+type Configuration struct {
+	// Opaque provides driver-specific configuration parameters.
+	//
+	// +required
```

Comment:
optional

Reply:
This is a bit odd at the moment: The field is optional, but validation will fail if it is not set (empty one-of).

I'll change to optional.

Reply:
Fixed.